scipy/optimize/_milp.py

```diff
@@ -0,0 +1,374 @@
+from ._optimize import OptimizeResult
+from ._linprog_highs import _highs_to_scipy_status_message
+
+
+def _constraints_to_components(constraints):
+    """
+    Convert sequence of constraints to a single set of components A, b_l, b_u.
+
+    `constraints` could be
+
+    1. A LinearConstraint
+    2. A tuple representing a LinearConstraint
+    3. An invalid object
+    4. A sequence of composed entirely of objects of type 1/2
+    5. A sequence containing at least one object of type 3
+
+    We want to accept 1, 2, and 4 and reject 3 and 5.
+    """
+    message = ("`constraints` (or each element within `constraints`) must be "
+               "convertible into an instance of "
+               "`scipy.optimize.LinearConstraint`.")
+    As = []
+    b_ls = []
+    b_us = []
+
+    # Accept case 1 by standardizing as case 4
+    if isinstance(constraints, LinearConstraint):
+        constraints = [constraints]
+    else:
+        # Reject case 3
+        try:
+            iter(constraints)
+        except TypeError as exc:
+            raise ValueError(message) from exc
+
+        # Accept case 2 by standardizing as case 4
+        if len(constraints) == 3:
+            # argument could be a single tuple representing a LinearConstraint
+            try:
+                constraints = [LinearConstraint(*constraints)]
+            except TypeError:
+                # argument was not a tuple representing a LinearConstraint
+                pass
+
+    # Address cases 4/5
+    for constraint in constraints:
+        # if it's not a LinearConstraint or something that represents a
+        # LinearConstraint at this point, it's invalid
+        if not isinstance(constraint, LinearConstraint):
+            try:
+                constraint = LinearConstraint(*constraint)
+            except TypeError as exc:
+                raise ValueError(message) from exc
+        As.append(csc_array(constraint.A))
+        b_ls.append(np.atleast_1d(constraint.lb).astype(np.double))
+        b_us.append(np.atleast_1d(constraint.ub).astype(np.double))
+
+    if len(As) > 1:
+        A = vstack(As)
+        b_l = np.concatenate(b_ls)
+        b_u = np.concatenate(b_us)
+    else:  # avoid unnecessary copying
+        A = As[0]
+        b_l = b_ls[0]
+        b_u = b_us[0]
+
+    return A, b_l, b_u
+
+
+def _milp_iv(c, integrality, bounds, constraints, options):
+    # objective IV
+    c = np.atleast_1d(c).astype(np.double)
+    if c.ndim != 1:
+        message = "`c` must be a one-dimensional array."
+        raise ValueError(message)
+
+    # integrality IV
+    message = ("`integrality` must contain integers 0-3 and be broadcastable "
+               "to `c.shape`.")
+    if integrality is None:
+        integrality = 0
+    try:
+        integrality = np.broadcast_to(integrality, c.shape).astype(np.uint8)
+    except ValueError:
+        raise ValueError(message)
+    if integrality.min() < 0 or integrality.max() > 3:
```

Comment:
Just found that this breaks on 0-sized `integrality` lists:
```python
  File "/home/nmckibben/Documents/scipy/scipy/optimize/_milp.py", line 351, in milp
    args_iv = _milp_iv(c, integrality, bounds, constraints, options)
  File "/home/nmckibben/Documents/scipy/scipy/optimize/_milp.py", line 91, in _milp_iv
    if integrality.min() < 0 or integrality.max() > 3:
  File "/venvs/scipy/lib/python3.8/site-packages/numpy/core/_methods.py", line 44, in _amin
    return umr_minimum(a, axis, None, out, keepdims, initial, where)
ValueError: zero-size array to reduction operation minimum which has no identity
```

(I haven't convinced myself how the broadcast above worked, but the fact I got the error surprised me)

Reply:
Maybe the real problem is when `c` is empty.
```
milp([])
milp([], integrality=1)
```
They both produce that error.

Otherwise, can you give an example? I can't find one, and I don't see how this could happen unless `c` is empty. As you said, `integrality` gets broadcast to the shape of `c`, which is at least 1d. I would think that the worst case would be like:
```
milp(np.array(1), integrality=np.array(1))
milp(1, integrality=1)
```
but there's no problem there.

Reply:
I ran into it while running the MIPLIB problems unattended and didn't save the logs, so I'm not sure what actually happened or what the problem parameters were.  I will restart running all the problems and see if I can find it again, but it seems like such a strange case that I'm not very invested in tracking it down.  But if it does show up again, I will ping this thread

Reply:
I made changes locally that should fix this but haven't pushed them. I'd like to wait to resolve the discussion about `status` before I push again.
```
-    if c.ndim != 1:
-        message = "`c` must be a one-dimensional array."
+    if c.ndim != 1 or c.size == 0 or not np.all(np.isfinite(c)):
+        message = ("`c` must be a one-dimensional array of finite numbers "
+                   "with at least one element.")
```

Reply:
@mckib2 Do you think this is resolved?

Reply:
Yes, we're good here